staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -919,6 +928,51 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 			}
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		// Update storage version records before returning the new handler.
+		// This means stale records can still exist in the cluster if an apiserver
+		// hasn't got any request. All apiservers have to at least serve one
+		// request (can be a GET) for a CR's storage version record to be
+		// up-to-date, and to trigger storage migration.
+		gr := schema.GroupResource{
+			Group:    crd.Spec.Group,
+			Resource: crd.Spec.Names.Plural,
+		}
+		encodingVersion := crd.Spec.Group + "/" + storageVersion
+		var decodableVersions []string
+		for _, v := range crd.Spec.Versions {
+			if !v.Served {
+				continue
+			}
+			decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+		}
+
+		appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+			ref := metav1.OwnerReference{
+				APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+				Kind:       "CustomResourceDefinition",
+				Name:       crd.Name,
+				UID:        crd.UID,
+			}
+			for _, r := range sv.OwnerReferences {
+				if r == ref {
+					return nil
+				}
+			}
+			sv.OwnerReferences = append(sv.OwnerReferences, ref)
```

Comment:
Append an owner reference to delete stale StorageVersions. This does not guarantee StorageVersions will be updated when the CRD gets updated.